keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,382 @@
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This feature enables the creation of a LoadBalancer Service that has different port definitions with different protocols. 
+
+## Motivation
+
+The ultimate goal of this feature is to support users that want to expose their applications via a single IP address but different L4 protocols with a cloud provided load-balancer. 
+The following issue and PR shows considerable interest from users that would benefit from this feature:
+https://github.com/kubernetes/kubernetes/issues/23880
+https://github.com/kubernetes/kubernetes/pull/75831
+
+### Goals
+
+The goals of this KEP are:
+
+- To analyze the impact of this feature with regard to current implementations of cloud-provider load-balancers
+- define how the activation of this feature could be made configurable if certain cloud-provider load-balancer implementations do not want to provide this feature
+
+### Non-Goals
+
+
+## Proposal
+
+The first thing proposed here is to relax the API validation in Kubernetes that currently rejects Service definitions with different protocols if their type is LoadBalancer. Kubernetes would not reject Service definitions like this from that point:
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: mixed-protocol
+spec:
+  type: LoadBalancer
+  ports:
+    - name: dns-udp
+      port: 53
+      protocol: UDP
+    - name: dns-tcp
+      port: 53
+      protocol: TCP
+  selector:
+    app: my-dns-server
+  ```
+
+Once that limit is removed those Service definitions will reach the Cloud Provider LB controller implementations. The logic of the particular Cloud Provider LB controller and of course the actual capabilities and architecture of the backing Cloud Load Balancer services determine how the actual exposure of the application really manifests. For this reason it is important to understand the capabilities of those backing services and to design this feature accordingly.
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes cluster user I want to expose an application that provides its service on different protocols with a single cloud provider load balancer IP. In order to achieve this I want to define different `ports` with mixed protocols in my Service definition of `type: LoadBalancer`
```

Comment:
This isn't a user story; it's just the KEP Summary with "As a user I want" prepended to it. :stuck_out_tongue: 

A user story should give a *concrete* example of something users want to do that they can't do currently, to give readers a solid understanding of why this KEP is (or maybe isn't) a good idea.

Eg, (totally making up a fake story based on the example in the Proposal section above), "I want to use Kubernetes to provide DNS service by exposing a LoadBalancer that does DNS over both TCP and UDP so that other DNS servers and hosts can make DNS requests to it regardless of whether they use TCP or UDP".

(I imagine you probably have some SCTP-based example of something users want to do with this?)

Reply:
Yeah. 
Even though I have some part in the SCTP support feature, and really nothing else in K8s, indeed, this time I have not started dealing with this KEP due to SCTP.
Updated the user stories (well, let's hope these can be called user stories now)